### PR TITLE
Add dnsmasq as a DNS cache in kube-dns pod

### PR DESCRIPTION
**Pull request by ArtfulCoder**
1. added dns-port flag
2. updated imports to not use the v1 version of the api objects.

### CODE REVIEW

**Comment by caesarxuchao** [comment]
LGTM. Thanks.

**Comment by ArtfulCoder** [comment]
@thockin

**Comment by ArtfulCoder** [comment]
Chao reviewed the import changes.
Waiting for Tim to review the DNS specific changes.

File: cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in [hunk 1]
```diff
@@ -54,13 +54,21 @@ spec:
         args:
         # command = "/kube-dns"
         - --domain={{ pillar['dns_domain'] }}.
+        - --dns-port=10053
         ports:
         - containerPort: 53
```
Comment (thockin) [inline cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in]: you should update the ports info to :10053 here and duplicate this into dnsmasq on :53

File: cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in [hunk 2]
```diff
@@ -54,13 +54,21 @@ spec:
...
           protocol: UDP
         - containerPort: 53
           name: dns-tcp
           protocol: TCP
+      - name: dnsmasq
+        image: gcr.io/google_containers/dnsmasq:1.0
+        command:
+        - dnsmasq
+        - -d
+        - -R
```
Comment (thockin) [inline cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in]: can you use long-flag names here?  --no-resolv and --no-daemon. Docs say --no-daemon is for debug, and to use --keep-in-foreground instead. 

Do we also need --no-hosts ? 

Do we need --cache-size=  here?  Even if it is just informational, we might eventually set it from a configmap

File: cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in [hunk 3]
```diff
@@ -54,13 +54,21 @@ spec:
...
+        - --dns-port=10053
         ports:
         - containerPort: 53
           name: dns
           protocol: UDP
         - containerPort: 53
           name: dns-tcp
           protocol: TCP
+      - name: dnsmasq
+        image: gcr.io/google_containers/dnsmasq:1.0
```
Comment (thockin) [inline cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in]: where is the dockerfile for this?

File: cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in (hunk 1 above)
Comment (ArtfulCoder) [inline cluster/saltbase/salt/kube-dns/kubedns-rc.yaml.in]: done.

**Comment by thockin** [comment]
LGTM - push your image and kick e2e

**Comment by ArtfulCoder** [comment]
@k8s-bot test this. issue:#IGNORE

**Comment by ArtfulCoder** [comment]
gce e2e never ran any tests and failed..
kicking it again

@k8s-bot test this issue: #IGNORE

**Comment by ArtfulCoder** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by ArtfulCoder** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by luxas** [comment]
@ArtfulCoder Is dnsmasq a required container to run in order to get it working with `1.1`?

**Comment by kjvalencik** [comment]
Would it be better to run `dnsmasq` in a separate RC / pod in production?

**Pull request merged**